OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,408 @@
+/* -------------------------------------------------------------------------- *
+ *                            OpenSim:  DataTable.h                           *
+ * -------------------------------------------------------------------------- *
+ * The OpenSim API is a toolkit for musculoskeletal modeling and simulation.  *
+ * See http://opensim.stanford.edu and the NOTICE file for more information.  *
+ * OpenSim is developed at Stanford University and supported by the US        *
+ * National Institutes of Health (U54 GM072970, R24 HD065690) and by DARPA    *
+ * through the Warrior Web program.                                           *
+ *                                                                            *
+ * Copyright (c) 2005-2015 Stanford University and the Authors                *
+ * Authors:                                                                   *
+ *                                                                            *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may    *
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+#ifndef OPENSIM_DATA_TABLE_H_
+#define OPENSIM_DATA_TABLE_H_
+
+/** \file
+This file defines the  DataTable_ class, which is used by OpenSim to provide an 
+in-memory container for data access and manipulation.                         */
+
+// Non-standard headers.
+#include "SimTKcommon.h"
+#include "OpenSim/Common/Exception.h"
+#include "OpenSim/Common/ValueArrayDictionary.h"
+
+namespace OpenSim {
+
+class InvalidRow : public Exception {};
+
+class IncorrectNumColumns : public InvalidRow {
+public:
+    IncorrectNumColumns(const std::string& file,
+                        size_t line,
+                        const std::string& func,
+                        size_t expected,
+                        size_t received) {
+        std::string msg{errorMessagePrefix(file, line, func)};
```

Comment:
I can move the general exceptions (I think only two -- IndexOutOfRange and KeyNotFound; rest are specific to DataTable).
I don't think it is just about adding a constructor to OpenSim::Exception. The goal (I think & @chrisdembia agrees) is to have a mechanism where the error message is formed through a recursive process -- every base class adds its own message to the message of the its derived class and OpenSim::Exception adds the file-name, line-number and function-name at the end of recursion. 
Following is an example message. Precise format of the message can be changed but the idea should be clear. A row is added using appendRow() and turns out to be invalid due to incorrect number of columns. The hierarchy is Exception <-- InvalidRow <-- IncorrectNumColumns.
```
/OpenSim/Common/DataTable.h:202: In function 'appendRow':
               attempting to add invalid row --  incorrect number of columns -- expected = 20 received = 22
```
One way to do this is to have each exception class call its direct base class's constructor with its message as an argument. This way by the time OpenSim::Exception constructor is called, the message is almost prepared except for -- file-name, line-number, function-name -- which OpenSim::Exception will add during its construction. 

After this point, there will still be redundancy in code at throw locations because the throw statement has to pass on the information about file-name, line-number and function-name. For this @chrisdembia  and I talked about having a macro replacing all the throw statements.

I thought doing all this was going to be part of another PR. If I don't do all of it now, I will have to come back to change things anyway. So let me know.